code/graph_algorithms/src/centroid_decomposition/centroid_decompostition.hpp

```diff
@@ -0,0 +1,80 @@
+#include <iostream>
+#include <vector>
+
+class Graph 
+{
+    int ver_;
+    std::vector<std::vector<int>> adj_;
+    std::vector<std::vector<int>> centroidTree_;
+    std::vector<int> sizes_;
+    std::vector<bool> marked_;
+    public:
+        Graph(int num) :
+            ver_(num),
+            adj_(num + 1),
+            centroidTree_(num + 1),
+            sizes_(num + 1),
+            marked_(num + 1)	
+            {}
+
+        void addEdge(int x, int y, std::vector<std::vector<int>> &graph);
+        void calcSizes(int cur, int par, std::vector<std::vector<int>> graph);
+        int findCentroid(int cur, int par, int vertices, std::vector<std::vector<int>> graph);
+        void decomposeTree(int cur, int par, int total, std::vector<std::vector<int>> graph);
+};
+
+void Graph::addEdge(int x, int y, std::vector<std::vector<int>> &graph)
+{
+    if (x < 0 || x > ver_)
+    {
+        throw std::out_of_range{"x is out of boundaries"};
+    }
+    if (y < 0 || y > ver_)
+    {
+        throw std::out_of_range{"y is out of boundaries"};
+    }
+    graph[x].push_back(y);
+}
+
+void Graph::calcSizes(int cur, int par, std::vector<std::vector<int>> graph)
+{
+    sizes_[cur] = 1;
+    for (const auto& to : graph[cur])
+    {
+        if (!(to == par || marked_[to] == true))
+        {
+            calcSizes (to, cur, graph);
```

Comment:
I hope this doesn't cause infinite recursion...

Reply:
No, it doesn't. Since it is a tree, every node has one parent. We check if our node is parent or not and if it isn't and not marked, it is a go.